test/sagemaker_tests/huggingface_pytorch/training/integration/sagemaker/test_neuronx.py

```diff
@@ -0,0 +1,113 @@
+
+# hyperparameters, which are passed into the training job
+hyperparameters = {
+    'model_name_or_path': 'hf-internal-testing/tiny-random-BertModel',
+    'dataset_name': 'philschmid/emotion',
+    'do_train': True,
+    'bf16': True,
+    'per_device_train_batch_size': 4,
+    'num_train_epochs': 1,
+    'output_dir': '/opt/ml/model'
+}
+
+# ValueError: Must setup local AWS configuration with a region supported by SageMaker.
+def retry_if_value_error(exception):
+    """Return True if we should retry (in this case when it's an ValueError), False otherwise"""
+    return isinstance(exception, ValueError)
+
+# TBD. This function is mainly there to handle capacity issues now. Once trn1 capacaity issues
+# are fixed, we can remove this function
+@retry(
+    stop_max_attempt_number=360, wait_fixed=10000, retry_on_exception=retry_if_value_error,
+)
+def invoke_neuron_helper_function(ecr_image, sagemaker_regions, helper_function, helper_function_args):
+    """
+    Used to invoke SM job defined in the helper functions in respective test file. The ECR image and the sagemaker
+    session are passed explicitly depending on the AWS region.
+    This function will rerun for all SM regions after a defined wait time if capacity issues are seen.
+
+    :param ecr_image: ECR image in us-west-2 region
+    :param sagemaker_regions: List of SageMaker regions
+    :param helper_function: Function to invoke
+    :param helper_function_args: Helper function args
+
+    :return: None
+    """
+    from ..... import get_ecr_image_region, get_sagemaker_session, get_ecr_image
+
+    ecr_image_region = get_ecr_image_region(ecr_image)
+    for region in sagemaker_regions:
+        sagemaker_session = get_sagemaker_session(region)
+        # Reupload the image to test region if needed
+        tested_ecr_image = get_ecr_image(ecr_image, region) if region != ecr_image_region else ecr_image
+        try:
+            helper_function(tested_ecr_image, sagemaker_session, **helper_function_args)
+            return
+        except sagemaker.exceptions.UnexpectedStatusException as e:
+            if "CapacityError" in str(e):
+                raise ValueError("CapacityError: Retry.")
+            else:
+                raise e
+
+@pytest.mark.processor("neuronx")
+@pytest.mark.model("hf-pt-qa-neuronx")
+@pytest.mark.neuronx_test
+@pytest.mark.skip_py2_containers
+def test_neuronx_question_answering(ecr_image, sagemaker_regions, py_version, instance_type):
+    function_args = {
+        'py_version': py_version,
+        'instance_type': instance_type,
+        'instance_count': 1,
+        'num_neuron_cores': 2,
+    }
+    invoke_neuron_helper_function(ecr_image, sagemaker_regions, _test_neuronx_question_answering_function, function_args)
+
+def _test_neuronx_question_answering_function(ecr_image, sagemaker_session, py_version, instance_type= "ml.trn1.32xlarge", instance_count=1, num_neuron_cores=2):
+    optimum_neuron_version = "0.0.3"
```

Comment:
Is it possible to make it future proof? For eg., can we just add a condition saying

```
if pytorch version is 1.13.0, then `optimum_neuron_version` = "0.0.3"

else "Error("optimum_neuron_version to be set for this framework version" )
```

If it seems to be more complicated, we can just add a comment here accordingly and add the condition as a follow up action item.

Reply:
Hi @trajanikant, 

The versioning of Optimum-neuron is more aligned with neuron SDK releases than PyTorch versions. So to say, for PyTorch 1.13.*, we can have multiple versions of Optimum-neuron. I am not so sure how to handle it elegantly as it is not a part of the buildspec. Currently, we need to add versioning of Optimum-neuron manually.

I added a conditional statement to make sure that for optimum-neuron 0.0.3 will be tested for PyTorch 1.13.*, and under other PyTorch versions, the test will be skipped. Not sure if that corresponds to the need.